vulnerabilities/overflow-underflow.md

```diff
@@ -1,73 +1,158 @@
+
+Consider the contract below:
 
 ```solidity
 uint8 public a = 255;
 
 function addition() public returns (uint8 result) {
     assembly {
         result := add(sload(a.slot), 1) // adding 1 will overflow and reset to 0
-        // using inline assembly
     }
 
     return result;
 }
 ```
-In the above code we are adding ``1`` into the variable with inline assembly and then returning the result. The variable result will overflow and 0 will be returned, despite this the contract will not throw an error or revert.
+
+In the above code we are adding 1 into the variable with inline assembly and then returning the result. The variable result will overflow and 0 will be returned, despite this the contract will NOT throw an error or revert!
+
+
+### Subtle Overflow with Smaller Integers (e.g., `uint128`)
+
+When using smaller integers like `uint128`, a more subtle overflow can occur because inline assembly operates with 256-bit values.
+
+Consider the contract below which provides a method to get a swap quote by adding one to the input amount:
+
+```solidity
+interface IDexPair {
+    function getSwapQuoteUint128(uint128 amountToSwap) external view returns(uint128);
+}
+
+contract DexPair is IDexPair {
+    function getSwapQuoteUint128(uint128 amountToSwap) external view returns(uint128 outputTokens) {
+        assembly {
+            outputTokens := add(amountToSwap, 1)
+            if lt(outputTokens, amountToSwap) { revert(0, 0) }
+        }
+    }
+}
+```
+
+- The problem is that the `add` opcode always produces a 256-bit result. For `uint128` maximum value (`type(uint128).max`), this will not overflow in the 256-bit space but will overflow when treated as `uint128`
+- The overflow check inside assembly fails because it checks against 256-bit values.
+
 
 ### Use of unchecked code block:
-Performing arithmetic operations inside the unchecked block saves a lot of gas because it omits several checks and opcodes. But, some of these opcodes are used in default arithmetic operations in 0.8 to check for underflow/overflow.
+Performing arithmetic operations inside the unchecked block saves a lot of gas because it omits several checks and opcodes. But some of these opcodes are used in default arithmetic operations in 0.8 to check for underflow/overflow.
 
 ```solidity
 uint8 public a = 255;
 
 function uncheck() public{
 
   unchecked {
-      a++;  // overflow and reset to 0 without reverting
+      a++;      // overflow and reset to 0 without reverting
   }
 
 }
 ```
 The unchecked code block is only recommended if you are sure that there is no possible way for the arithmetic to overflow or underflow.
 
+
+### Mitigation Strategies
+
+1. Using `addmod`
+
+```solidity
+function getSwapQuoteUint128(uint128 amountToSwap) external view returns(uint128 outputTokens) {
+    assembly {
+        outputTokens := addmod(amountToSwap, 1, 340282366920938463463374607431768211455)
```

Comment:
In most circumstances, we'd rather have execution revert rather than actually overflowing the output. In fact, I can't really think of any circumstances where we wouldn't

Reply:
Yeah I don't think we should suggest allowing the overflow